### ISSUE TITLE
Memory on acceptance tests on ember 1.13.10+

### ISSUE THREAD

**Issue by calderas** [emberjs/ember.js#12618]
test repo: https://github.com/calderas/leaky-mcleakerson

After upgrading our app to ember 1.13 chrome crashes when running Acceptance Tests

```
V8 error: Allocation failed - process out of memory (CALL_AND_RETRY_LAST).  Current memory usage: 1949 MB
```

Following the same approach as https://github.com/emberjs/ember.js/issues/11748 I ran some tests on a simple app to compare different versions of ember

### Observations: 
* Memory drops when running unit tests on ember 1.12.
* Memory allocated for tests is  ~ twice on ember 1.13 & 2.1 for same app.
* Memory allocation tends to increase overtime on acceptance tests.

### Questions:
* Is this expected ?
* Thoughts on how to debug further ?

### Memory comparison

![screen shot 2015-11-16 at 12 59 31 pm](https://cloud.githubusercontent.com/assets/118375/11191170/d0f1f530-8c67-11e5-8248-3a14047bb550.png)
![screen shot 2015-11-16 at 12 58 20 pm](https://cloud.githubusercontent.com/assets/118375/11191169/d0ef5668-8c67-11e5-9b95-ceea14cfab28.png)
![screen shot 2015-11-16 at 12 58 49 pm](https://cloud.githubusercontent.com/assets/118375/11191171/d0f3e822-8c67-11e5-9e16-86639734783b.png)

Looking at the timeline seems like event listeners remain after tests have completed and detached elements show up on heap profiles.

<img width="1432" alt="ef067318-892c-11e5-8a4d-fe75297533cb" src="https://cloud.githubusercontent.com/assets/118375/11191592/0638b89e-8c6a-11e5-9cfd-4e4a9579ead2.png">
<img width="1267" alt="f09d0494-892c-11e5-8c0f-aba694f2b135" src="https://cloud.githubusercontent.com/assets/118375/11191591/06376eee-8c6a-11e5-8bb0-739752603ab6.png">

test repo: https://github.com/calderas/leaky-mcleakerson

**Comment by jcope2013** [comment]
this one seems related also https://github.com/emberjs/ember.js/issues/12490 and I have been having similar issues with my app

**Comment by lmcardle** [comment]
I've got the same thing going on. I've been trying to track the issue down for the past few weeks to no avail. I am also on 1.13.10, here is what my timeline and profile look like.
<img width="1195" alt="memory timeline" src="https://cloud.githubusercontent.com/assets/186442/11413273/271aa5e8-939c-11e5-9781-777e9e47a9ff.png">
<img width="996" alt="memory profile" src="https://cloud.githubusercontent.com/assets/186442/11413275/2d636fd4-939c-11e5-9f8b-09ef68f16f67.png">

**Comment by lmcardle** [comment]
I've also been struggling with this on 1.13.10. No idea if this would be helpful, but here is my timeline and heap profile.
<img width="996" alt="memory profile" src="https://cloud.githubusercontent.com/assets/186442/11416232/8bc8f200-93c0-11e5-8fb5-ff46d14ed75f.png">
<img width="1195" alt="memory timeline" src="https://cloud.githubusercontent.com/assets/186442/11416233/8bcc2218-93c0-11e5-9685-743cb9bcdb6b.png">

**Comment by stefanpenner** [comment]
@lmcardle it is unfortunately not at all helpful, the only thing that is valuable is actual examples demonstrating leaks.

**Comment by stefanpenner** [comment]
> Thoughts on how to debug further ?

yes, heap snapshot, find the retainers and address them.

**Comment by stefanpenner** [comment]
I took a look, although there may be more leaks. I believe I found a v8 bug. It appears to be leaking due to an inlined function preserving its context...

After some digging, I found:

<img width="580" alt="screen shot 2015-11-26 at 12 08 50 am" src="https://cloud.githubusercontent.com/assets/1377/11427539/76352730-9418-11e5-8eb9-77302131b7ba.png">

Which seems to indicate the `normalize` function (which doesn't close over anything) via its code and related deopt info, is retaining the resolver.

The following stack appears inlined (causing the grief).

1. [resolve.normalize](https://github.com/emberjs/ember.js/blob/fb2ae6f41aaddcf14a9c96b245553b1f56c971fd/packages/ember-application/lib/system/application.js#L1451-L1455)
2. [this.registry.normalizeFullName = this.applicationRegistry.normalizeFullName;
](https://github.com/emberjs/ember.js/blob/v1.13.11/packages/ember-application/lib/system/application-instance.js#L97)
3. [Registry.prototypr.normalize](https://github.com/emberjs/ember.js/blob/v1.13.11/packages/container/lib/registry.js#L353-L357)


I then pinged @mraleph on twitter, who recommended i disabling inlining and try again.

```
/Applications/Google\ Chrome\ Canary.app/Contents/MacOS/Google\ Chrome\ Canary --js-flags="--nouse_inlining" --user-data-dir=/tmp/foobar
```

Manually preventing this specific set of functions from inlining also prevents the leak. [hacky patch](https://github.com/emberjs/ember.js/pull/12652/files)

And the container leak i noticed went away...

**Comment by stefanpenner** [comment]
FYI: when searching for memory leaks, having examples that create excessive noise only creates distractions and make finding the source of the leak more difficult.

The key is finding the minimal setup required to cause a leak, basically deleting parts of the example until the leak stops, and following that process until only a concise example remain.

[After deleting nearly everything](https://github.com/stefanpenner/leaker/commit/5d709749633359868b29e9a559c5dc89e1093a50) the leak was still obvious, and made for a much easier time debugging. Ultimate the leak appears to be a v8 bug, but several work-arounds are possible. With one of those work-arounds, full re-instating the provided tests also no longer leak containers.

**Comment by rwjblue** [comment]
This should have been addressed by #12666, but I'm reopening so that we can confirm.

**Comment by shen6653** [comment]
Is this fix available for 1.13?

**Comment by stefanpenner** [comment]
> Is this fix available for 1.13?

@shen6653 I don't know if it has been backported, we would likely accept a PR doing so.

**Comment by mixonic** [comment]
It seems like this just needs confirmation against the testing repo provided in the original ticket.

**Comment by oso-mate** [comment]
Forked the original leaker repo and upgraded to Ember 2.3.2. The memory leak issue seems to have been fixed in the latest version.

<img width="1440" alt="screen shot 2016-03-28 at 12 23 58 pm" src="https://cloud.githubusercontent.com/assets/318347/14088130/9d0d1a4a-f4e2-11e5-9ec1-1f15987a5691.png">

**Comment by stefanpenner** [comment]
> The memory leak issue seems to have been fixed for this version as part of: #12666

to be pedantic, we are working around a v8 bug in that PR :trollface:

**Comment by calderas** [comment]
I just tested this again with the latest ember/ember-cli and looks so much better !!! 
https://github.com/calderas/leaky-mcleakerson/tree/feature/ember-2.4.3

Memory doesn't increase between acceptance tests anymore
<img width="602" alt="screen shot 2016-03-28 at 6 27 29 pm" src="https://cloud.githubusercontent.com/assets/118375/14092245/932cf348-f513-11e5-8a35-95d89f41d3be.png">

Thanks for the constant improvements !

**Comment by acorncom** [comment]
Per comments above, it looks like this there have been improvements.  I'll close this for now, but please comment if we should re-open it.

**Comment by micky2be** [comment]
I would love to see a fix done for 1.13.x

I'm currently updating an app (quit big) from 1.12.2 to 2.x, and passing by 1.13 to fix all deprecation warning is a most.
So if we cannot run the tests of our app, it's gonna be very hard to know if any refactoring has been done correctly.

**Comment by Kuzirashi** [comment]
Is there any way to manually clear memory between acceptance tests run? I'm also interested in this. I doubt any fix will be backported to 1.13 and it's better to upgrade to Ember 2, but maybe there's some workaround?

**Comment by micky2be** [comment]
@Kuzirashi You could do like me.
I run multiple group of tests one by one with different filter.
It reloads PhantomJS, with fresh memory.
I've been surprise that by doing that it was faster than keeping the same instance of PahntomJS
